itest/mint_fund_seal_test.go

```diff
@@ -0,0 +1,706 @@
+package itest
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcec/v2/schnorr"
+	"github.com/btcsuite/btcd/btcutil/psbt"
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	taprootassets "github.com/lightninglabs/taproot-assets"
+	"github.com/lightninglabs/taproot-assets/asset"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/internal/test"
+	"github.com/lightninglabs/taproot-assets/tappsbt"
+	"github.com/lightninglabs/taproot-assets/taprpc"
+	wrpc "github.com/lightninglabs/taproot-assets/taprpc/assetwalletrpc"
+	"github.com/lightninglabs/taproot-assets/taprpc/mintrpc"
+	"github.com/lightninglabs/taproot-assets/taprpc/universerpc"
+	"github.com/lightninglabs/taproot-assets/tapsend"
+	"github.com/lightninglabs/taproot-assets/universe"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/keychain"
+	"github.com/stretchr/testify/require"
+)
+
+// testMintFundSealAssets tests that we're able to mint assets with custom
+// script keys, group keys, and group key tapscript roots. Once minted, we
+// also want to reissue assets with a custom asset group witness, and spend
+// an asset with a custom script key witness. An asset group with a tapscript
+// root should also be imported into a new node correctly.
+func testMintFundSealAssets(t *harnessTest) {
+	// We create a second tapd node that will be used to simulate a second
+	// party in the test. This tapd node is connected to lnd "Bob".
+	var (
+		aliceTapd = t.tapd
+		aliceLnd  = t.lndHarness.Alice
+		bobLnd    = t.lndHarness.Bob
+		bobTapd   = setupTapdHarness(t.t, t, bobLnd, t.universeServer)
+	)
+
+	aliceLndClient, err := t.newLndClient(aliceLnd)
+	require.NoError(t.t, err)
+
+	defer func() {
+		require.NoError(t.t, bobTapd.stop(!*noDelete))
+	}()
+
+	displayGroupWitness := func(wit *taprpc.GroupWitness) string {
+		return fmt.Sprintf(
+			"genID: %x\nWitness: %x\n",
+			wit.GenesisId, wit.Witness,
+		)
+	}
+
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultWaitTimeout)
+	defer cancel()
+
+	aliceLndKeyRing := taprootassets.NewLndRpcKeyRing(
+		&aliceLndClient.LndServices,
+	)
+
+	// Derive a random key on an LND node, with a key family not matching
+	// the Taproot Assets key family.
+	deriveRandomKey := func(
+		keyRing *taprootassets.LndRpcKeyRing) keychain.KeyDescriptor {
+
+		var (
+			randFam = test.RandInt31n(math.MaxInt32)
+			randInd = test.RandInt31n(255)
+			desc    keychain.KeyDescriptor
+			err     error
+		)
+
+		// Ensure that we use a different key family from tapd.
+		for randFam == asset.TaprootAssetsKeyFamily {
+			randFam = test.RandInt31n(math.MaxInt32)
+		}
+
+		// Use DeriveNextKey to reach our desired key index.
+		for i := int32(0); i != randInd; i++ {
+			desc, err = keyRing.DeriveNextKey(
```

Comment:
micro-nit: in case we already have a non-zero amount of keys for this family, we could set `i = desc.Index` to short cut and don't derive too many keys.

Reply:
Would require a change to the condition though, so would need to be `i <= randInt`, otherwise it might turn into an endless loop.

Reply:
Good catch! I think using the random int as an offset from the current index should work.